File: anvil/swift/Anvil/Sources/Anvil/AnvilSemanticColors.swift
import SwiftUI

/// `--color-black-static`: true black in every tier.
private let blackStatic = Color(rgb: 0x000000)

/// Anvil's semantic color layer, ported from `src/theme.css`.
///
/// These are the brand-meaningful colors (brand fills, info/success/warning/
/// critical/promo surfaces, foregrounds, borders, focus rings) built on top of
/// the primitive palettes. The layer is exposed to apps because brand semantics
/// such as "success surface" or "warning border" have no SwiftUI equivalent.
///
/// All values are deliberately faithful to the web CSS and match the Android
/// (`AnvilSemanticColors.kt`), Flutter (`anvil_semantic_colors.dart`) and
/// React Native (`semantic.ts`) ports field-for-field. Note that in the dark
/// tiers Anvil flips `black`/`white` and reverses the ramps, exactly as the
/// CSS does.
public struct AnvilSemanticColors: Sendable, Equatable {
    // Surfaces
    public let background: Color
    public let surface: Color
    public let surfaceHovered: Color
    public let surfaceSecondary: Color
    public let overlay: Color
    public let overlayHovered: Color
    public let inverse: Color
    // Brand + control fills
    public let fillBrandPrimary: Color
    public let fillBrandPrimaryHovered: Color
    public let fillBrandSecondary: Color
    public let fillBrandSecondaryHovered: Color
    public let fillDisabled: Color
    public let fillOpaque: Color
    public let fillControlSelected: Color
    // Status fills
    public let fillInfo: Color
    public let fillInfoSecondary: Color
    public let fillSuccess: Color
    public let fillSuccessSecondary: Color
    public let fillWarning: Color
    public let fillWarningSecondary: Color
    public let fillCritical: Color
    public let fillCriticalSecondary: Color
    public let fillPromo: Color
    public let fillPromoSecondary: Color
    public let fillPromoSecondaryHovered: Color
    // Foregrounds
    public let foregroundNormal: Color
    public let foregroundSubtle: Color
    public let foregroundAccent: Color
    public let foregroundWarning: Color
    public let foregroundDisabled: Color
    public let foregroundOnFill: Color
    public let foregroundOnInverse: Color
    public let foregroundOnBrandPrimary: Color
    // Borders
    public let borderNormal: Color
    public let borderStrong: Color
    public let borderInfo: Color
    public let borderSuccess: Color
    public let borderWarning: Color
    public let borderCritical: Color
    public let borderPromo: Color
    public let borderControlSelected: Color
    // Focus / progress
    public let ring: Color
    public let focusOutline: Color
    public let progress: Color
    public let progressRemaining: Color
    // Code highlighting
    public let codeBackground: Color
    public let codeForeground: Color
    public let codeMuted: Color
    public let codeKeyword: Color
    public let codeLiteral: Color
    public let codeString: Color
    public let codeSelection: Color
}

/// Anvil light theme semantic colors (`:root.light` in theme.css).
public func lightAnvilColors() -> AnvilSemanticColors {
    let p = AnvilLightPalette.self
    return AnvilSemanticColors(
        background: p.neutral50,
        surface: p.white,
        surfaceHovered: p.neutral500.withAlpha(0.02),
        surfaceSecondary: p.neutral500.withAlpha(0.04),
        overlay: p.white,
        overlayHovered: p.neutral50,
        inverse: p.neutral950,

        // Web light tier: brand primary is black-on-white (readability);
        // citrus brand fills remain in the dark/high-contrast tiers.
        fillBrandPrimary: p.neutral950,
        fillBrandPrimaryHovered: p.neutral950,
        fillBrandSecondary: p.neutral950,
        fillBrandSecondaryHovered: p.neutral950,
        fillDisabled: p.neutral300,
        fillOpaque: p.neutral950.withAlpha(0.03),
        fillControlSelected: p.black,

        fillInfo: p.sky500,
        fillInfoSecondary: p.sky50,
        fillSuccess: p.emerald600,
        fillSuccessSecondary: p.emerald100,
        fillWarning: p.orange600,
        fillWarningSecondary: p.orange100,
        fillCritical: p.red600,
        fillCriticalSecondary: p.red100,
        fillPromo: p.violet500,
        fillPromoSecondary: p.violet100,
        fillPromoSecondaryHovered: p.violet200,

        foregroundNormal: p.neutral950,
        foregroundSubtle: p.neutral600,
        foregroundAccent: p.citrus700,
        foregroundWarning: p.orange600,
        foregroundDisabled: p.neutral500,
        foregroundOnFill: p.white,
        foregroundOnInverse: p.white,
        foregroundOnBrandPrimary: p.white,

        borderNormal: p.neutral950.withAlpha(0.12),
        borderStrong: p.neutral950,
        borderInfo: p.sky500,
        borderSuccess: p.emerald600,
        borderWarning: p.orange600,
        borderCritical: p.red600,
        borderPromo: p.violet500,
        borderControlSelected: p.black,

        ring: p.citrus700,
        focusOutline: p.citrus700,
        progress: p.citrus700,
        progressRemaining: p.citrus700.withAlpha(0.20),

        codeBackground: p.white,
        codeForeground: p.neutral950,
        codeMuted: p.neutral600,
        codeKeyword: p.citrus700,
        codeLiteral: p.sunset600,
        codeString: p.sky800,
        codeSelection: p.citrus500.withAlpha(0.15)
    )
}

/// Anvil dark theme semantic colors (`:root.dark` in theme.css).
public func darkAnvilColors() -> AnvilSemanticColors {
    let p = AnvilDarkPalette.self
    return AnvilSemanticColors(
        background: p.neutral100,
        surface: p.neutral200,
        surfaceHovered: p.neutral500.withAlpha(0.07),
        surfaceSecondary: p.neutral500.withAlpha(0.15),
        overlay: p.neutral200,
        overlayHovered: p.neutral100,
        inverse: p.neutral950,

        fillBrandPrimary: p.citrus600,
        fillBrandPrimaryHovered: p.citrus700,
        fillBrandSecondary: p.neutral950,
        fillBrandSecondaryHovered: p.neutral950,
        fillDisabled: p.neutral600,
        fillOpaque: p.neutral950.withAlpha(0.03),
        fillControlSelected: p.primary,

        fillInfo: p.sky600,
        fillInfoSecondary: p.sky50,
        fillSuccess: p.emerald600,
        fillSuccessSecondary: p.emerald50,
        fillWarning: p.orange600,
        fillWarningSecondary: p.orange50,
        fillCritical: p.red600,
        fillCriticalSecondary: p.red50,
        fillPromo: p.violet600,
        fillPromoSecondary: p.violet50,
        fillPromoSecondaryHovered: p.violet100,

        foregroundNormal: p.neutral950,
        foregroundSubtle: p.neutral700,
        foregroundAccent: p.citrus600,
        foregroundWarning: p.orange600,
        foregroundDisabled: p.neutral500,
        // `white`/`black` are flipped in the dark tier, matching the CSS.
        foregroundOnFill: p.white,
        foregroundOnInverse: p.white,
        foregroundOnBrandPrimary: blackStatic,

        borderNormal: p.neutral950.withAlpha(0.25),
        borderStrong: p.neutral950,
        borderInfo: p.sky500,
        borderSuccess: p.emerald600,
        borderWarning: p.orange600,
        borderCritical: p.red600,
        borderPromo: p.violet500,
        borderControlSelected: p.primary,

        ring: p.citrus700,
        focusOutline: p.citrus600,
        // The bare `:root` block of theme.css pins --progress to citrus-700 for
        // every tier (the dark block does not override it).
        progress: p.citrus700,
        progressRemaining: p.citrus700.withAlpha(0.20),

        codeBackground: p.neutral200,
        codeForeground: p.neutral950,
        codeMuted: p.neutral700,
        codeKeyword: p.citrus700,
        codeLiteral: p.sunset600,
        codeString: p.sky800,
        codeSelection: p.citrus500.withAlpha(0.15)
    )
}

/// Anvil light high-contrast semantic colors (`:root.light-high-contrast`).
public func lightHighContrastAnvilColors() -> AnvilSemanticColors {
    let p = AnvilLightHighContrastPalette.self
    return AnvilSemanticColors(
        background: p.neutral50,
        surface: p.white,
        surfaceHovered: p.neutral500.withAlpha(0.02),
        surfaceSecondary: p.neutral500.withAlpha(0.04),
        overlay: p.white,
        overlayHovered: p.neutral50,
        inverse: p.neutral950,

        fillBrandPrimary: p.citrus800,
        fillBrandPrimaryHovered: p.citrus900,
        fillBrandSecondary: p.neutral950,
        fillBrandSecondaryHovered: p.neutral950,
        fillDisabled: p.neutral300.withAlpha(0.60),
        fillOpaque: p.neutral950.withAlpha(0.03),
        fillControlSelected: p.black,

        fillInfo: p.sky500,
        fillInfoSecondary: p.sky50,
        fillSuccess: p.emerald600,
        fillSuccessSecondary: p.emerald100,
        fillWarning: p.orange600,
        fillWarningSecondary: p.orange100,
        fillCritical: p.red800,
        fillCriticalSecondary: p.red100,
        fillPromo: p.violet500,
        fillPromoSecondary: p.violet100,
        fillPromoSecondaryHovered: p.violet200,

        foregroundNormal: p.black,
        foregroundSubtle: p.neutral800,
        foregroundAccent: p.citrus700,
        foregroundWarning: p.orange600,
        foregroundDisabled: p.neutral500,
        foregroundOnFill: p.white,
        foregroundOnInverse: p.white,
        foregroundOnBrandPrimary: blackStatic,

        borderNormal: p.neutral950,
        borderStrong: p.neutral950,
        borderInfo: p.sky500,
        borderSuccess: p.emerald600,
        borderWarning: p.orange600,
        borderCritical: p.red600,
        borderPromo: p.violet500,
        borderControlSelected: p.black,

        ring: p.citrus700,
        focusOutline: p.citrus700,
        progress: p.citrus800,
        progressRemaining: p.citrus800.withAlpha(0.20),

        codeBackground: p.white,
        codeForeground: p.black,
        codeMuted: p.neutral800,
        codeKeyword: p.citrus900,
        codeLiteral: p.sunset800,
        codeString: p.sky950,
        codeSelection: p.citrus500.withAlpha(0.15)
    )
}

/// Anvil dark high-contrast semantic colors (`:root.dark-high-contrast`).
///
/// Cascade note: the web applies `dark-high-contrast` as a *single* class —
/// the `:root.dark` override block does NOT apply. So this tier = `@theme`
/// defaults + the bare `:root` overrides + the dark-high-contrast block.
public func darkHighContrastAnvilColors() -> AnvilSemanticColors {
    let p = AnvilDarkHighContrastPalette.self
    return AnvilSemanticColors(
        background: p.neutral50,
        surface: p.white, // flipped in the dark tier — resolves to black
        surfaceHovered: p.neutral500.withAlpha(0.02),
        surfaceSecondary: p.neutral500.withAlpha(0.04),
        overlay: p.white, // flipped — black
        overlayHovered: p.neutral50,
        inverse: p.neutral950,

        fillBrandPrimary: p.citrus800,
        fillBrandPrimaryHovered: p.citrus900,
        fillBrandSecondary: p.neutral950,
        fillBrandSecondaryHovered: p.neutral950,
        fillDisabled: p.black, // per CSS --fill-disabled: var(--color-black)
        fillOpaque: p.neutral950.withAlpha(0.03),
        fillControlSelected: p.primary,

        fillInfo: p.sky500,
        fillInfoSecondary: p.sky50,
        fillSuccess: p.emerald600,
        fillSuccessSecondary: p.emerald50,
        fillWarning: p.orange600,
        fillWarningSecondary: p.orange50,
        fillCritical: p.red800,
        fillCriticalSecondary: p.red50,
        fillPromo: p.violet500,
        fillPromoSecondary: p.violet50,
        fillPromoSecondaryHovered: p.violet100,

        foregroundNormal: p.black, // flipped — resolves to white
        foregroundSubtle: p.neutral800,
        foregroundAccent: p.citrus800,
        foregroundWarning: p.orange600,
        foregroundDisabled: p.neutral500,
        foregroundOnFill: p.black, // flipped — resolves to white
        foregroundOnInverse: p.white, // flipped — resolves to black
        foregroundOnBrandPrimary: blackStatic,

        borderNormal: p.neutral950,
        borderStrong: p.neutral950,
        borderInfo: p.sky800,
        borderSuccess: p.emerald800,
        borderWarning: p.orange800,
        borderCritical: p.red800,
        borderPromo: p.violet800,
        borderControlSelected: p.primary,

        ring: p.citrus700,
        focusOutline: p.citrus700,
        progress: p.citrus700,
        progressRemaining: p.citrus700.withAlpha(0.20),

        codeBackground: p.white, // flipped — black
        codeForeground: p.black,
        codeMuted: p.neutral800,
        codeKeyword: p.citrus900,
        codeLiteral: p.sunset800,
        codeString: p.sky950,
        codeSelection: p.citrus500.withAlpha(0.15)
    )
}

/// Resolves semantic colors for a given [AnvilThemeTier].
public func anvilSemanticColors(tier: AnvilThemeTier) -> AnvilSemanticColors {
    switch tier {
    case .light: return lightAnvilColors()
    case .dark: return darkAnvilColors()
    case .lightHighContrast: return lightHighContrastAnvilColors()
    case .darkHighContrast: return darkHighContrastAnvilColors()
    }
}

File: anvil/android/anvil-tokens/src/commonMain/kotlin/live/ditto/anvil/tokens/AnvilSemanticColors.kt
package live.ditto.anvil.tokens

import androidx.compose.ui.graphics.Color

/** `--color-black-static`: true black in every tier. */
private val BlackStatic = Color(0xFF000000)

/**
 * Anvil's semantic color layer, ported from `src/theme.css`.
 *
 * These are the brand-meaningful colors (brand fills, info/success/warning/
 * critical/promo surfaces, foregrounds, borders, focus rings) built on top of
 * the primitive palettes ([AnvilLightPalette], [AnvilDarkPalette] and the
 * high-contrast variants). The Material `ColorScheme` in the theme modules is
 * derived from this layer, and the layer itself is exposed to apps because
 * brand semantics such as "success surface" or "warning border" have no
 * Material role.
 *
 * All values are deliberately faithful to the web CSS. The light tier uses a
 * black brand fill with white foreground (readability); Anvil's signature
 * black-on-citrus brand fill remains in the dark and high-contrast tiers
 * (accessed via [foregroundOnBrandPrimary]). Note that in the dark tiers Anvil
 * flips `black`/`white` and reverses the ramps, exactly as the CSS does.
 */
data class AnvilSemanticColors(
    // Surfaces
    val background: Color,
    val surface: Color,
    val surfaceHovered: Color,
    val surfaceSecondary: Color,
    val overlay: Color,
    val overlayHovered: Color,
    val inverse: Color,
    // Brand + control fills
    val fillBrandPrimary: Color,
    val fillBrandPrimaryHovered: Color,
    val fillBrandSecondary: Color,
    val fillBrandSecondaryHovered: Color,
    val fillDisabled: Color,
    val fillOpaque: Color,
    val fillControlSelected: Color,
    // Status fills
    val fillInfo: Color,
    val fillInfoSecondary: Color,
    val fillSuccess: Color,
    val fillSuccessSecondary: Color,
    val fillWarning: Color,
    val fillWarningSecondary: Color,
    val fillCritical: Color,
    val fillCriticalSecondary: Color,
    val fillPromo: Color,
    val fillPromoSecondary: Color,
    val fillPromoSecondaryHovered: Color,
    // Foregrounds
    val foregroundNormal: Color,
    val foregroundSubtle: Color,
    val foregroundAccent: Color,
    val foregroundWarning: Color,
    val foregroundDisabled: Color,
    val foregroundOnFill: Color,
    val foregroundOnInverse: Color,
    val foregroundOnBrandPrimary: Color,
    // Borders
    val borderNormal: Color,
    val borderStrong: Color,
    val borderInfo: Color,
    val borderSuccess: Color,
    val borderWarning: Color,
    val borderCritical: Color,
    val borderPromo: Color,
    val borderControlSelected: Color,
    // Focus / progress
    val ring: Color,
    val focusOutline: Color,
    val progress: Color,
    val progressRemaining: Color,
    // Code highlighting
    val codeBackground: Color,
    val codeForeground: Color,
    val codeMuted: Color,
    val codeKeyword: Color,
    val codeLiteral: Color,
    val codeString: Color,
    val codeSelection: Color,
)

/** Anvil light theme semantic colors (`:root.light` in theme.css). */
fun lightAnvilColors(): AnvilSemanticColors = with(AnvilLightPalette) {
    AnvilSemanticColors(
        background = neutral50,
        surface = white,
        surfaceHovered = neutral500.copy(alpha = 0.02f),
        surfaceSecondary = neutral500.copy(alpha = 0.04f),
        overlay = white,
        overlayHovered = neutral50,
        inverse = neutral950,

        // Web light tier: brand primary is black-on-white (readability);
        // citrus brand fills remain in the dark/high-contrast tiers.
        fillBrandPrimary = neutral950,
        fillBrandPrimaryHovered = neutral950,
        fillBrandSecondary = neutral950,
        fillBrandSecondaryHovered = neutral950,
        fillDisabled = neutral300,
        fillOpaque = neutral950.copy(alpha = 0.03f),
        fillControlSelected = black,

        fillInfo = sky500,
        fillInfoSecondary = sky50,
        fillSuccess = emerald600,
        fillSuccessSecondary = emerald100,
        fillWarning = orange600,
        fillWarningSecondary = orange100,
        fillCritical = red600,
        fillCriticalSecondary = red100,
        fillPromo = violet500,
        fillPromoSecondary = violet100,
        fillPromoSecondaryHovered = violet200,

        foregroundNormal = neutral950,
        foregroundSubtle = neutral600,
        foregroundAccent = citrus700,
        foregroundWarning = orange600,
        foregroundDisabled = neutral500,
        foregroundOnFill = white,
        foregroundOnInverse = white,
        foregroundOnBrandPrimary = white,

        borderNormal = neutral950.copy(alpha = 0.12f),
        borderStrong = neutral950,
        borderInfo = sky500,
        borderSuccess = emerald600,
        borderWarning = orange600,
        borderCritical = red600,
        borderPromo = violet500,
        borderControlSelected = black,

        ring = citrus700,
        focusOutline = citrus700,
        progress = citrus700,
        progressRemaining = citrus700.copy(alpha = 0.20f),

        codeBackground = white,
        codeForeground = neutral950,
        codeMuted = neutral600,
        codeKeyword = citrus700,
        codeLiteral = sunset600,
        codeString = sky800,
        codeSelection = citrus500.copy(alpha = 0.15f),
    )
}

/** Anvil dark theme semantic colors (`:root.dark` in theme.css). */
fun darkAnvilColors(): AnvilSemanticColors = with(AnvilDarkPalette) {
    AnvilSemanticColors(
        background = neutral100,
        surface = neutral200,
        surfaceHovered = neutral500.copy(alpha = 0.07f),
        surfaceSecondary = neutral500.copy(alpha = 0.15f),
        overlay = neutral200,
        overlayHovered = neutral100,
        inverse = neutral950,

        fillBrandPrimary = citrus600,
        fillBrandPrimaryHovered = citrus700,
        fillBrandSecondary = neutral950,
        fillBrandSecondaryHovered = neutral950,
        fillDisabled = neutral600,
        fillOpaque = neutral950.copy(alpha = 0.03f),
        fillControlSelected = primary,

        fillInfo = sky600,
        fillInfoSecondary = sky50,
        fillSuccess = emerald600,
        fillSuccessSecondary = emerald50,
        fillWarning = orange600,
        fillWarningSecondary = orange50,
        fillCritical = red600,
        fillCriticalSecondary = red50,
        fillPromo = violet600,
        fillPromoSecondary = violet50,
        fillPromoSecondaryHovered = violet100,

        foregroundNormal = neutral950,
        foregroundSubtle = neutral700,
        foregroundAccent = citrus600,
        foregroundWarning = orange600,
        foregroundDisabled = neutral500,
        // `white`/`black` are flipped in the dark tier, matching the CSS.
        foregroundOnFill = white,
        foregroundOnInverse = white,
        foregroundOnBrandPrimary = BlackStatic,

        borderNormal = neutral950.copy(alpha = 0.25f),
        borderStrong = neutral950,
        borderInfo = sky500,
        borderSuccess = emerald600,
        borderWarning = orange600,
        borderCritical = red600,
        borderPromo = violet500,
        borderControlSelected = primary,

        ring = citrus700,
        focusOutline = citrus600,
        // The bare `:root` block of theme.css pins --progress to citrus-700 for
        // every tier (the dark block does not override it).
        progress = citrus700,
        progressRemaining = citrus700.copy(alpha = 0.20f),

        codeBackground = neutral200,
        codeForeground = neutral950,
        codeMuted = neutral700,
        codeKeyword = citrus700,
        codeLiteral = sunset600,
        codeString = sky800,
        codeSelection = citrus500.copy(alpha = 0.15f),
    )
}

/** Anvil light high-contrast semantic colors (`:root.light-high-contrast`). */
fun lightHighContrastAnvilColors(): AnvilSemanticColors = with(AnvilLightHighContrastPalette) {
    AnvilSemanticColors(
        background = neutral50,
        surface = white,
        surfaceHovered = neutral500.copy(alpha = 0.02f),
        surfaceSecondary = neutral500.copy(alpha = 0.04f),
        overlay = white,
        overlayHovered = neutral50,
        inverse = neutral950,

        fillBrandPrimary = citrus800,
        fillBrandPrimaryHovered = citrus900,
        fillBrandSecondary = neutral950,
        fillBrandSecondaryHovered = neutral950,
        fillDisabled = neutral300.copy(alpha = 0.60f),
        fillOpaque = neutral950.copy(alpha = 0.03f),
        fillControlSelected = black,

        fillInfo = sky500,
        fillInfoSecondary = sky50,
        fillSuccess = emerald600,
        fillSuccessSecondary = emerald100,
        fillWarning = orange600,
        fillWarningSecondary = orange100,
        fillCritical = red800,
        fillCriticalSecondary = red100,
        fillPromo = violet500,
        fillPromoSecondary = violet100,
        fillPromoSecondaryHovered = violet200,

        foregroundNormal = black,
        foregroundSubtle = neutral800,
        foregroundAccent = citrus700,
        foregroundWarning = orange600,
        foregroundDisabled = neutral500,
        foregroundOnFill = white,
        foregroundOnInverse = white,
        foregroundOnBrandPrimary = BlackStatic,

        borderNormal = neutral950,
        borderStrong = neutral950,
        borderInfo = sky500,
        borderSuccess = emerald600,
        borderWarning = orange600,
        borderCritical = red600,
        borderPromo = violet500,
        borderControlSelected = black,

        ring = citrus700,
        focusOutline = citrus700,
        progress = citrus800,
        progressRemaining = citrus800.copy(alpha = 0.20f),

        codeBackground = white,
        codeForeground = black,
        codeMuted = neutral800,
        codeKeyword = citrus900,
        codeLiteral = sunset800,
        codeString = sky950,
        codeSelection = citrus500.copy(alpha = 0.15f),
    )
}

/**
 * Anvil dark high-contrast semantic colors (`:root.dark-high-contrast`).
 *
 * Cascade note: the web applies `dark-high-contrast` as a *single* class —
 * the `:root.dark` override block does NOT apply. So this tier = `@theme`
 * defaults + the bare `:root` overrides (shadows, progress, code) + the
 * dark-high-contrast block. Notably: surfaces stay at the dark-HC defaults
 * (white→black flip, 2%/4% hover alpha), and `fillInfo`/`fillPromo` keep the
 * default 500 steps.
 */
fun darkHighContrastAnvilColors(): AnvilSemanticColors = with(AnvilDarkHighContrastPalette) {
    AnvilSemanticColors(
        background = neutral50,
        surface = white, // flipped in the dark tier — resolves to black
        surfaceHovered = neutral500.copy(alpha = 0.02f),
        surfaceSecondary = neutral500.copy(alpha = 0.04f),
        overlay = white, // flipped — black
        overlayHovered = neutral50,
        inverse = neutral950,

        fillBrandPrimary = citrus800,
        fillBrandPrimaryHovered = citrus900,
        fillBrandSecondary = neutral950,
        fillBrandSecondaryHovered = neutral950,
        fillDisabled = black, // per CSS --fill-disabled: var(--color-black)
        fillOpaque = neutral950.copy(alpha = 0.03f),
        fillControlSelected = primary,

        fillInfo = sky500,
        fillInfoSecondary = sky50,
        fillSuccess = emerald600,
        fillSuccessSecondary = emerald50,
        fillWarning = orange600,
        fillWarningSecondary = orange50,
        fillCritical = red800,
        fillCriticalSecondary = red50,
        fillPromo = violet500,
        fillPromoSecondary = violet50,
        fillPromoSecondaryHovered = violet100,

        foregroundNormal = black, // flipped in the dark tier — resolves to white
        foregroundSubtle = neutral800,
        foregroundAccent = citrus800,
        foregroundWarning = orange600,
        foregroundDisabled = neutral500,
        foregroundOnFill = black, // flipped — resolves to white
        foregroundOnInverse = white, // flipped — resolves to black
        foregroundOnBrandPrimary = BlackStatic,

        borderNormal = neutral950,
        borderStrong = neutral950,
        borderInfo = sky800,
        borderSuccess = emerald800,
        borderWarning = orange800,
        borderCritical = red800,
        borderPromo = violet800,
        borderControlSelected = primary,

        ring = citrus700,
        focusOutline = citrus700,
        progress = citrus700,
        progressRemaining = citrus700.copy(alpha = 0.20f),

        codeBackground = white, // = --background-surface, flipped — black
        codeForeground = black,
        codeMuted = neutral800,
        codeKeyword = citrus900,
        codeLiteral = sunset800,
        codeString = sky950,
        codeSelection = citrus500.copy(alpha = 0.15f),
    )
}

/** Resolves semantic colors for a given [AnvilThemeTier]. */
fun anvilSemanticColors(tier: AnvilThemeTier): AnvilSemanticColors =
    when (tier) {
        AnvilThemeTier.Light -> lightAnvilColors()
        AnvilThemeTier.Dark -> darkAnvilColors()
        AnvilThemeTier.LightHighContrast -> lightHighContrastAnvilColors()
        AnvilThemeTier.DarkHighContrast -> darkHighContrastAnvilColors()
    }
